pkg/profile/artifact.go

```diff
@@ -37,19 +34,62 @@ func (p *Profile) CreateArtifacts(ctx context.Context) error {
 		if !ok {
 			return fmt.Errorf("object %v cannot be asserted to client.Object", o)
 		}
-		p.log.Info("creating...", "kind", obj.GetObjectKind().GroupVersionKind().Kind, "resource", obj.GetName())
-		if err := p.client.Create(ctx, obj); err != nil {
-			return fmt.Errorf("failed to create %s: %w", obj.GetObjectKind().GroupVersionKind().Kind, err)
+		if err := p.reconcileArtifact(obj); err != nil {
+			return err
 		}
 	}
 
 	p.log.Info("all artifacts created")
 	return nil
 }
+func (p *Profile) reconcileArtifact(desiredObj client.Object) error {
+	existingObj := desiredObj.DeepCopyObject().(client.Object)
+	err := p.client.Get(p.ctx, client.ObjectKeyFromObject(desiredObj), existingObj)
+	if err == nil {
+		return p.updateResource(existingObj, desiredObj)
+	} else if apierrors.IsNotFound(err) {
+		p.log.Info("creating...", "kind", desiredObj.GetObjectKind().GroupVersionKind().Kind, "resource", desiredObj.GetName())
+		if err := p.client.Create(p.ctx, desiredObj); err != nil {
+			return fmt.Errorf("failed to create %s: %w", desiredObj.GetObjectKind().GroupVersionKind().Kind, err)
+		}
+		return nil
+	} else {
+		return fmt.Errorf("failed to get resource %s %s/%s: %w", desiredObj.GetObjectKind().GroupVersionKind().Kind, desiredObj.GetNamespace(), desiredObj.GetName(), err)
+	}
+}
+
+func (p *Profile) updateResource(existingObj, desiredObj client.Object) error {
+	switch desiredObj := desiredObj.(type) {
+	case *sourcev1.GitRepository:
+		if !GitRepoRequiresUpdate(existingObj.(*sourcev1.GitRepository), desiredObj) {
+			return nil
+		}
+		existingObj.(*sourcev1.GitRepository).Spec = desiredObj.Spec
+	case *sourcev1.HelmRepository:
+		if !HelmRepoRequiresUpdate(existingObj.(*sourcev1.HelmRepository), desiredObj) {
```

Comment:
It feels like `HelmRepoRequiresUpdate` is only exported for the tests, it's not used outside of this package?

Reply:
yes, that is true. I don't like testing private functions, but at the same time I felt it was important to have tests for the update checking. Not sure what to do tbh

Reply:
If it was me, I wouldn't put those tests in the `_test` package, that's used for testing the external interface of a package, and you're testing the internal behaviour here?

Reply:
What's wrong with testing private functions? Unit tests can test whatever they want, can't they? I mean, I see that we mostly test out of package with `_test`, but I feel like exporting this function for the sake of testing is also a smell in itself. I think. :)

Reply:
> If it was me, I wouldn't put those tests in the _test package, that's used for testing the external interface of a package, and you're testing the internal behaviour here?

I'm not sure where to put them then :thinking: I really dislike testing private functions, but having these tests are important (while the checking logic checks individual fields, if we can do some other `deepequal` its less of a problem). The only other alternative I see is testing all of these cases through the public function they are used in, which means bloating out the `ReconcileArtifact` tests even more.

We could make the func private, and then use the `export_test.go` pattern to export is publicly for the tests only, but its still the same problem of testing a private func, just not making this func public for no reason.

Reply:
I don't think we need the `export_test` pattern either, just move the tests into the package, you're not testing the external behaviour.